### ISSUE TITLE
Masked scatter update

### ISSUE THREAD

**Issue by awni** [ml-explore/mlx#2149]
It would be very convenient to have a scatter update which takes a boolean mask. This is a common use case in for example VLMs where you update one vector of embeddings with another array of image embeddings with the update locations based on a mask.

Something like:

```
a[mask] = b
```

That might be complex if we wanted to allow mixing boolean and non-boolean indices. But maybe it's fine to allow only a single array index if it's boolean.

Or maybe more like the torch API [masked_scatter](https://pytorch.org/docs/stable/generated/torch.Tensor.masked_scatter.html)

**Comment by romanoneg** [comment]
Hi! I'm somewhat new to this project (have been using it for a while but want to contribute) and would love to work on this! I've checked out the contribution guidelines and I'm up-to-date with the library. However I want to make sure I don't go against predefined guidelines for the project so I have a couple of questions about this specific implementation.

Firstly, which api would be better? `a[mask] = b` or `mx.mast_scatter()` or both? 

Otherwise, this functionality would be fairly easily implemented with a nonzero function (working on bools) but it seems that, from past github issues, that nonzero function wouldn't be possible (https://github.com/ml-explore/mlx/issues/1641). Is there a reason why this wouldn't work if we were to create a `mx.nonzero` that takes a `count` as input and would return a `number_of_nonzero_elements` as output? Something like this:
```py
a = mx.array([1,0,0,2],[0,3,4,0])
out, num_nz = mx.nonzero(a, count=8)
a_nz = a[tuple(out[:, :num_nz.item()])]

print(out) 
# out -> [[0,0,1,1,0,0,0,0], [0,3,1,2,0,0,0,0]] outputs a [D,count] array where D is the num of dimensions of a

print(num_nz) 
# num_nz -> 4

print(a_nz) 
# a_nz -> [1,2,3,4]
```
and this nonzero function would return only the first `count` elements if there were more than `count`, returning `num_nz = count`.

If this is definitely not the way you want to go about this `masked_scatter` implementation, would a workaround using a flurry of `flattens`, `cumsums`, and `wheres` be preferred?

**Comment by awni** [comment]
Having a nonzero which takes a count is not unreasonable. It's basically the [API Jax provides](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.nonzero.html).

I don't quite follow how you would use that to implement the masked scatter. Maybe you can explain what you have in mind.

> would a workaround using a flurry of flattens, cumsums, and wheres be preferred

I would probably start with that.. and then we can optimize it with something more custom if needed.

**Comment by romanoneg** [comment]
> I don't quite follow how you would use that to implement the masked scatter. Maybe you can explain what you have in mind.

If I understand what needs to get done right, then I think it should just be something along the lines of (in pseudocode):
```py
def masked_scatter(a, mask, b):
    indices = mx.nonzero(mask)
    return mx.scatter(a=a, indices=indices, updates=b) 
```
to simplify the pseudocode skipping over the count for the nonzero, the needed format of indices, and the axes in scatter.
do note: This would only work as long as nonzero also works with bools.

**Comment by awni** [comment]
I think the problem with that approach is `nonzero` requires a count (and pad token).. and the masked scatter shouldn't require a count to work.

**Comment by romanoneg** [comment]
Right good point, I had thought about this for a bit and completely forgotten about it. There is the trick of just passing the count as the number of elements in the mask, but that can get expensive on the memory. I'll work on the implementation using the other primitives and maybe try to see if its a worthwhile tradeoff using this trick later.

**Comment by awni** [comment]
> There is the trick of just passing the count as the number of elements in the mask, but that can get expensive on the memory.

Right aside from being inefficient, It's also not easy to do it that way. The mask can have fewer than `count` nonzero elements. But you always return a mask with size `count` regardless of the number of nonzeros. But the scatter should only take the nonzero indices (which would require a graph evaluation).

**Comment by CC-Yeh** [comment]
Hey @romanoneg,

For the GPU path, I first compute `idx_mask` with `cumsum`, then use `idx_mask` + `mask` to do the assignment in a second kernel. This adds extra memory overhead `O(|mask|)` for `idx_mask`, but the performance looks strong in #2663.
